lib/spack/spack/modules/lmod.py

```diff
@@ -0,0 +1,414 @@
+#
+# This file is part of Spack.
+# Created by Todd Gamblin, tgamblin@llnl.gov, All rights reserved.
+# LLNL-CODE-647188
+#
+# For details, see https://github.com/llnl/spack
+# Please also see the LICENSE file for our notice and the LGPL.
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License (as
+# published by the Free Software Foundation) version 2.1, February 1999.
+#
+# This program is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the IMPLIED WARRANTY OF
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the terms and
+# conditions of the GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+##############################################################################
+
+import os.path
+import llnl.util.lang as lang
+import spack.compilers
+import spack.spec
+import spack.error
+import itertools
+import collections
+import spack.tengine as tengine
+
+from .common import BaseConfiguration, BaseFileLayout
+from .common import BaseContext, BaseModuleFileWriter, configuration
+
+#: LMOD specific part of the configuration
+configuration = configuration.get('lmod', {})
+
+#: Caches the configuration {spec_hash: configuration}
+configuration_registry = {}
+
+
+def make_configuration(spec):
+    """Returns the lmod configuration for spec"""
+    key = spec.dag_hash()
+    try:
+        return configuration_registry[key]
+    except KeyError:
+        return configuration_registry.setdefault(key, LmodConfiguration(spec))
+
+
+def make_layout(spec):
+    """Returns the layout information for spec """
+    conf = make_configuration(spec)
+    return LmodFileLayout(conf)
+
+
+def make_context(spec):
+    """Returns the context information for spec"""
+    conf = make_configuration(spec)
+    return LmodContext(conf)
+
+
+class LmodConfiguration(BaseConfiguration):
+    """Configuration class for lmod module files."""
+
+    @property
+    def core_compilers(self):
+        """Returns the list of "Core" compilers
+
+        :raises CoreCompilersNotFoundError: if the key was not
+            specified in the configuration file or the sequence
+            is empty
+        """
+        value = configuration.get('core_compilers')
+        if value is None:
+            msg = '\'core_compilers\' key not found in configuration file'
+            raise CoreCompilersNotFoundError(msg)
+        if not value:
+            msg = '\'core_compilers\' list cannot be empty'
+            raise CoreCompilersNotFoundError(msg)
+        return value
+
+    @property
+    def hierarchy_tokens(self):
+        """Returns the list of tokens that are part of the modulefile
+        hierarchy. 'compiler' is always present.
```

Comment:
Maybe I worded the sentence in an unclear way. What I mean is that at least you get back the compiler part. Because that can be inferred from the spec and **needs** to be part of the hierarchy.